SofaKernel/modules/SofaBaseCollision/src/SofaBaseCollision/BruteForceDetection.cpp

```diff
@@ -71,76 +69,114 @@ void BruteForceDetection::addCollisionModel(core::CollisionModel *cm)
+        cmPairs.emplace_back(cm, cm);
     }
 
+    core::CollisionModel* finalCollisionModel = cm->getLast();
+
     // Browse all other collision models to check if there is a potential collision (conservative check)
-    for (sofa::helper::vector<core::CollisionModel*>::iterator it = collisionModels.begin(); it != collisionModels.end(); ++it)
+    for (auto* cm2 : collisionModels)
     {
-        core::CollisionModel* cm2 = *it;
-
         // ignore this pair if both are NOT simulated (inactive)
         if (!cm->isSimulated() && !cm2->isSimulated())
         {
             continue;
         }
 
-        if (!keepCollisionBetween(cm->getLast(), cm2->getLast()))
+        if (!keepCollisionBetween(finalCollisionModel, cm2->getLast()))
             continue;
 
         bool swapModels = false;
         core::collision::ElementIntersector* intersector = intersectionMethod->findIntersector(cm, cm2, swapModels);
         if (intersector == nullptr)
             continue;
 
-        core::CollisionModel* cm1 = (swapModels?cm2:cm);
-        cm2 = (swapModels?cm:cm2);
+        core::CollisionModel* cm1 = cm;
+        if (swapModels)
+        {
+            std::swap(cm1, cm2);
+        }
 
         // Here we assume a single root element is present in both models
         if (intersector->canIntersect(cm1->begin(), cm2->begin()))
         {
-            cmPairs.push_back(std::make_pair(cm1, cm2));
+            //both collision models will be further examined in the narrow phase
+            cmPairs.emplace_back(cm1, cm2);
         }
     }
+
+    //accumulate CollisionModel's in a vector so the next CollisionModel can be tested against all previous ones
     collisionModels.push_back(cm);
 }
 
+bool BruteForceDetection::intersectWithBoxModel(core::CollisionModel *cm) const
+{
+    bool swapModels = false;
+    core::collision::ElementIntersector* intersector = intersectionMethod->findIntersector(cm, boxModel.get(), swapModels);
+    if (intersector)
+    {
+        core::CollisionModel* cm1 = (swapModels?boxModel.get():cm);
+        core::CollisionModel* cm2 = (swapModels?cm:boxModel.get());
+
+        // Here we assume a single root element is present in both models
+        return intersector->canIntersect(cm1->begin(), cm2->begin());
+    }
+
+    return true;
+}
+
+bool BruteForceDetection::doesSelfCollide(core::CollisionModel *cm) const
+{
+    if (cm->isSimulated() && cm->getLast()->canCollideWith(cm->getLast()))
+    {
+        // self collision
+        bool swapModels = false;
+        core::collision::ElementIntersector* intersector = intersectionMethod->findIntersector(cm, cm, swapModels);
+        if (intersector != nullptr)
+        {
+            return intersector->canIntersect(cm->begin(), cm->begin());
+        }
+    }
+
+    return false;
+}
+
+bool BruteForceDetection::isSelfCollision(core::CollisionModel* cm1, core::CollisionModel* cm2)
+{
+    return (cm1->getContext() == cm2->getContext());
```

Comment:
Should we rely on the context to determine the fact there are on the same object ?

Reply:
In BruteForceDetection, the broad phase checks if an object must be tested against itself:
https://github.com/sofa-framework/sofa/blob/339241a4a36e8d4ca0f44107421f32c58dcd8540/SofaKernel/modules/SofaBaseCollision/src/SofaBaseCollision/BruteForceDetection.cpp#L81
, based on the value of `bSelfCollision`:
https://github.com/sofa-framework/sofa/blob/4026a23b534056afe180b1d250fbc4302feadb07/SofaKernel/modules/SofaCore/src/sofa/core/CollisionModel.cpp#L167

If the object must be tested against itself, then `cm1==cm2` in the function you mention. So I guess, the context is also the same.
If `cm1!=cm2`, I don't know if they both can have the same context.

In any case, I copied the code which was initially here:
https://github.com/sofa-framework/sofa/blob/4026a23b534056afe180b1d250fbc4302feadb07/SofaKernel/modules/SofaBaseCollision/src/SofaBaseCollision/BruteForceDetection.cpp#L169

Reply:
Ah I did not see that it was the same before.
I have some doubt about the test of "the same context" but if it was the same as before, this is not really the scope of the PR.
(and selfCollision in SOFA is... not really here, lets say 😅)